### ISSUE TITLE
Retire or rebuild the legacy project/block data layer before reactivation

### ISSUE THREAD

**Issue by teamleaderleo** [teamleaderleo/scrapbook#269]
## Current disposition

The old `/dashboard` project/block system is not the current product path: `app/dashboard/layout.tsx` redirects the entire legacy dashboard to `/space`.

Do not spend effort making its Drizzle query prettier until repository and deployed-build evidence answers whether the legacy routes, client components, and Server Actions are still reachable or bundled. The first decision is **retire, quarantine, or rebuild**, not query-style cleanup.

### Authority is caller supplied

Legacy client components import a public constant:

```ts
export const ADMIN_UUID = '410544b2-4001-4271-9855-fec4b6a6442a';
```

They pass that value into Server Actions such as `deleteProject`, `createBlock`, and `addBlockToProject`. The actions accept `accountId` as an argument rather than deriving ownership from authenticated server state.

A UUID is an identifier, not authorization.

### Link writes do not prove ownership

`addBlockToProject(blockId, projectId, accountId)` inserts directly into `project_block_link` without first proving that both the block and project belong to the supplied account.

`createBlockInProject` creates a block and then links it through two separate operations rather than one transaction. A failed link can leave an orphan block.

### Read joins trust inconsistent ownership rows

Project queries filter the root project by account, but join:

- project/block links by project ID;
- blocks by block ID;
- tag associations by associated ID and entity type;
- tags by tag ID.

They do not also bind the joined rows to the same account at each ownership boundary. The schema repeats `accountId` on projects, blocks, links, tags, and associations, but does not enforce composite ownership relationships tying those records together.

If inconsistent rows exist, aggregation can attach a foreign block or tag to an otherwise authorized root record.

### Query shape multiplies rows

The project query joins project tags and project blocks in one relation before aggregating both with `DISTINCT`. A project with several tags and blocks produces a tag × block intermediate set. The deeper variant also executes a correlated subquery for every aggregated block to collect block tags.

This may be acceptable for tiny data, but it should not be optimized without representative cardinalities and an execution plan.

### The layer contains substantial dormant residue

The current tree includes:

- fully commented hooks;
- alternate basic/full fetch functions;
- hand-written JSON aggregation;
- debug logging of query/results;
- legacy dashboard forms and components;
- React `cache()` wrappers with unclear invalidation semantics;
- a public portfolio note claiming optimized CRUD performance from this older system.

The repository should not preserve this indefinitely as if it were current production architecture.

## Phase 1 — reachability and data decision

Before changing behavior:

1. inspect the production build’s Server Action manifest and route output;
2. verify whether any non-dashboard route imports the legacy actions, data functions, schema, or components;
3. verify whether deployed endpoints can invoke the actions despite the dashboard redirect;
4. inventory whether production data still exists in the related tables and whether it has retention value;
5. identify whether `/space` or any planned feature needs the old project/block model;
6. classify every legacy file as active, migration input, portfolio evidence, or removable residue.

Use no production mutations during the audit.

## Decision A — retire

Prefer retirement when no current product flow depends on the layer.

A retirement plan must:

- preserve or export any retained data before schema removal;
- remove legacy routes, client components, actions, data functions, constants, and unused dependencies in reviewable slices;
- verify Server Action and route manifests no longer expose the old surface;
- update portfolio copy so historical claims remain accurate without implying the code is current;
- decide separately whether old migrations remain as history or move to an archive.

## Decision B — quarantine

Use quarantine only when data must remain accessible during migration.

- make every legacy mutation fail closed by default;
- derive identity from authenticated server state, never a client argument;
- require an explicit migration/admin capability for any retained access;
- keep the dashboard redirect;
- add no new product dependency on the old schema.

## Decision C — rebuild

Rebuild only when a current `/space` or product requirement genuinely needs relational projects, blocks, and tags.

### Ownership invariant

Every read and write must prove one server-derived principal owns every participating row.

At minimum:

- project, block, tag, link, and association ownership are checked in the same transaction;
- schema constraints prevent cross-account project/block links and tag associations;
- unique constraints prevent duplicate links and entity-type collisions;
- authorization is not represented by `ADMIN_UUID` in client code;
- foreign-key and deletion behavior is explicit;
- rows affected are checked before reporting success.

### Query contracts

Define named projections instead of one universal nested fetch:

- project list summary;
- project detail;
- project with ordered block references;
- block detail with projects/tags;
- explicit admin/export projection.

For each projection record:

```text
caller and route
required columns and relations
ordering and pagination
maximum expected cardinality
query count and execution plan
cache scope and invalidation
not-found and unauthorized behavior
```

Prefer separate bounded queries or Drizzle relations when they reduce cross products and improve type ownership. Raw SQL aggregation is acceptable only when it wins against a measured alternative and has focused tests.

## Required regression matrix

- caller cannot select another account through an argument;
- cross-account project/block link is rejected at write time and by the database;
- cross-account tag association is rejected;
- inconsistent seeded rows cannot leak through reads;
- create-and-link is atomic;
- delete/update reports not-found or unauthorized when zero rows match;
- project list does not load block bodies unless requested;
- deterministic ordering exists for projects, blocks, and tags;
- pagination does not duplicate or omit records;
- representative tag/block cardinalities have retained query plans;
- caches do not survive relevant mutations incorrectly.

## Stopping rule

Finish with one evidenced disposition:

- legacy layer removed;
- legacy layer quarantined with a migration deadline;
- current requirements justify a rebuilt, server-authorized data layer;
- retained data blocks action, with the exact export or migration requirement named.

Do not migrate databases or adopt Convex merely to avoid making this ownership decision. Database-provider selection remains tracked separately in #348.

Tracks research-inbox triage #497.

**Comment by MikelBai** [comment]
I don’t think this is that hard but I’m just really sleepy right now and there’s a lot of code to update

**Comment by teamleaderleo** [comment]
Brownfield reconciliation on 2026-08-15: the retirement disposition described by this issue has already been carried through and current main no longer contains the risky legacy product surface.

Completed slices:

- PR #511 / `1cc7cb3`: removed the complete `/dashboard` App Router tree and added coverage that former dashboard URLs 404 while `/space` remains active;
- PR #512 / `4c6fc4b`: removed the orphaned project/block/tag Server Actions, query modules, caches, dashboard controls, Zustand stores, and legacy navigation (1,128 source lines removed), explicitly preserving the useful architectural lessons on this issue;
- PR #521 / `436435a`: removed the remaining disconnected dashboard/form/suggestion UI and pruned 31 unused direct dependencies, with one consolidated deprecation-boundary contract.

Current repository search finds no `ADMIN_UUID`, `deleteProject`, `createBlockInProject`, or `addBlockToProject`, and `app/dashboard` is absent. The issue's ownership/query analysis remains valuable historical guidance, while the executable legacy layer it warned about has been retired. Closing as completed.
